sno/structure.py

```diff
@@ -604,239 +620,116 @@ def diff(self, other, ds_filter=UNFILTERED, reverse=False):
-                    to_write = [(name, json_pack(delta.new.value))]
-                else:
-                    to_write = [(name, ensure_bytes(delta.new.value))]
-                for path, data in to_write:
-                    blob_id = repo.create_blob(data)
-                    meta_tree = git_util.replace_subtree(repo, meta_tree, path, blob_id)
+            if delta.type == "update" and old_path not in self.tree:
+                conflicts = True
+                click.echo(
+                    f"{self.path}: Trying to update nonexistent feature: {old_key}"
+                )
+                continue
 
-        orig_tree = git_util.replace_subtree(repo, orig_tree, meta_path, meta_tree)
+            if delta.type == "update" and not self._features_equal(
+                self.get_feature(old_key), delta.old_value, geom_column_name
+            ):
+                conflicts = True
+                click.echo(
+                    f"{self.path}: Trying to update already-changed feature: {old_key}"
+                )
+                continue
 
-        geom_column_name = self.geom_column_name
-        deltas_by_directory = {}
-        for delta in dataset_diff.get("feature", {}).values():
-            if delta.type == "delete":
-                pks = {delta.old.value[pk_field]}
-            elif delta.type == "insert":
-                pks = {delta.new.value[pk_field]}
-            elif delta.type == "update":
-                pks = {delta.old.value[pk_field], delta.new.value[pk_field]}
-
-            for pk in pks:
-                feature_path = self.encode_1pk_to_path(pk)
-                pieces = feature_path.rsplit('/')
-                dir_ = deltas_by_directory
-                for piece in pieces[:-1]:
-                    dir_ = dir_.setdefault(piece, {})
-                dir_[pieces[-1]] = delta
-
-        new_tree, conflicts = self._recursive_build_feature_tree(
-            orig_tree=orig_tree,
-            repo=repo,
-            deltas_dict=deltas_by_directory,
-            encode_kwargs=encode_kwargs,
-            pk_field=pk_field,
-            geom_column_name=geom_column_name,
-        )
+            # Actually write the feature diff:
+            if old_path and old_path != new_path:
+                tree_builder.remove(old_path)
+            if delta.new_value:
+                path, data = self.encode_feature(
+                    delta.new.value, relative=True, **encode_kwargs
+                )
+                tree_builder.insert(path, data)
 
         if conflicts:
             raise InvalidOperation(
                 "Patch does not apply", exit_code=PATCH_DOES_NOT_APPLY,
             )
-        return new_tree
+
+    def _features_equal(self, lhs, rhs, geom_column_name):
+        # FIXME: actually compare the geometries here.
+        # Turns out this is quite hard - geometries are hard to compare sanely.
+        # Even if we add hacks to ignore endianness, WKB seems to vary a bit,
```

Comment:
So in terms of _our_ geometries, our WKB encoding should (now) be deterministic. But, there is also:
* [`ST_OrderingEquals()`](https://postgis.net/docs/ST_OrderingEquals.html) — this is `OGRGeometry.Equals(other)` (and `OGRGeometry.Equal(other)` is a legacy API to the same behaviour).
* [`ST_Equals()`](https://postgis.net/docs/ST_Equals.html) — afaict this isn't implemented in GDAL, but is in the underlying GEOS library

Would one of them do what you want?

Reply:
Won't do in this in this change - this logic predates this change. I'll look into this for a follow up change